docs/source/speechlm2/datasets.rst

```diff
@@ -0,0 +1,406 @@
+        "recording": {
+            "id": "conversation_1_user",
+            "path": "/path/to/audio/conversation_1_user.wav",
+            "sampling_rate": 16000,
+            "num_samples": 171200,
+            "duration": 10.7
+        },
+        "custom": {
+            "target_audio": {
+                "id": "conversation_1_assistant",
+                "path": "/path/to/audio/conversation_1_assistant.wav",
+                "sampling_rate": 22050,
+                "num_samples": 235935,
+                "duration": 10.7
+            }
+        }
+    }
+
+The DuplexS2SDataset performs several key operations when processing data:
+
+1. **Turn Identification**: Each cut contains a list of `supervisions` with objects of type `lhotse.SupervisionSegment` that represent conversation turns with corresponding text and speaker information.
+
+2. **Speaker Role Separation**: The text of each supervision is tokenized and identified as the model's output (when `supervision.speaker` is in `output_roles`, e.g., "agent" or "Assistant") or the model's input (when in `input_roles`, e.g., "user" or "User").
+
+3. **Token Sequence Generation**:
+   - `target_tokens` and `source_tokens` arrays are created with a length equal to `lhotse.utils.compute_num_frames(cut.duration, frame_length, cut.sampling_rate)`
+   - The `frame_length` parameter (typically 80ms) determines the temporal resolution of token assignments
+   - Each token is assigned to a position based on its corresponding audio segment's timing
+
+4. **Token Offset Calculation**:
+   - The starting position for each turn's tokens is determined using `lhotse.utils.compute_num_frames(supervision.start, frame_length)`
+   - This ensures tokens are aligned with their corresponding audio segments
+
+5. **Length Validation**:
+   - If token sequences are too long compared to the audio duration, warnings are emitted
+   - Tokens that extend beyond the audio length are truncated
+
+This process ensures that the model can correctly align audio input with corresponding text, and learn to generate appropriate responses based on the conversation context.
+
+DuplexS2SDataset
+****************
+
+This dataset class is designed for models that handle both speech understanding and speech generation. It processes audio inputs and prepares them for the model along with corresponding text.
+
+.. code-block:: python
+
+    from nemo.collections.speechlm2.data import DuplexS2SDataset
+    
+    dataset = DuplexS2SDataset(
+        tokenizer=model.tokenizer,                   # Text tokenizer
+        frame_length=0.08,                          # Frame length in seconds
+        source_sample_rate=16000,                   # Input audio sample rate
+        target_sample_rate=22050,                   # Output audio sample rate
+        input_roles=["user", "User"],               # Roles considered as input
+        output_roles=["agent", "Assistant"]         # Roles considered as output
+    )
+
+SALMDataset Structure
```

Comment:
If possible, can we add an explanation for `audio_locator_tag`?

Reply:
Does this work?

```
Each audio turn is represented as a single token, defined in ``audio_locator_tag`` property, and automatically added to the model's tokenizer inside model code.
This token is replaced during the training/generation pass with its corresponding audio segment representation.
```